src/openvic-simulation/utility/TslHelper.hpp

```diff
@@ -0,0 +1,128 @@
+#pragma once
+
+#include <type_traits>
+
+#include <tsl/ordered_map.h>
+#include <tsl/ordered_set.h>
+
+#include "openvic-simulation/utility/Utility.hpp"
+
+namespace OpenVic {
+	template<typename Key, typename Value, typename Map = tsl::ordered_map<Key, Value>>
+	struct _OrderedMapMutable {
+		using map_type = Map;
+		struct ordered_iterator {
+			using key_type = Key;
+			using mapped_type = Value;
+			using pair_type = std::pair<Key const&, Value&>;
+			using value_type = pair_type;
+			using iterator = typename map_type::iterator;
+
+			using iterator_category = std::random_access_iterator_tag;
+			using difference_type = typename map_type::values_container_type::iterator::difference_type;
+
+			pair_type operator*() {
+				return { m_iterator.key(), m_iterator.value() };
+			}
+
+			ordered_iterator& operator++() {
+				++m_iterator;
+				return *this;
+			}
+			ordered_iterator& operator--() {
+				--m_iterator;
+				return *this;
+			}
+
+			ordered_iterator operator++(int) {
+				ordered_iterator tmp(*this);
+				++(*this);
+				return tmp;
+			}
+			ordered_iterator operator--(int) {
+				ordered_iterator tmp(*this);
+				--(*this);
+				return tmp;
+			}
+
+			pair_type operator[](difference_type n) {
+				auto tmp = m_iterator[n];
+				return { tmp.key(), tmp.value() };
+			}
```

Comment:
value is necessary to modify the value by reference, returning the references as a pair, it should be fine so long as the map's elements aren't moved or removed...

Reply:
I've not been experiencing this, but also mutable_iterator only really exists to be a forward and backward iterator, I never honestly intended it to work for anything else. (so range for loops are trivial to work with) I'm not sure why this would become invalidated some of the time.

Reply:
This is still not working for me:
- `tsl::detail_ordered_hash::ordered_hash::ordered_iterator::operator[]` returns `const std::pair<key_type, mapped_type>&`, the contents of which is copied into `auto` which becomes `std::pair<key_type, mapped_type>`. This can be verified by adding these `static_assert`s between lines 50 and 51:
```c++
static_assert(std::is_same_v<const std::pair<key_type, mapped_type>&, decltype(m_iterator[n])>);
static_assert(std::is_same_v<std::pair<key_type, mapped_type>, decltype(tmp)>);
```
- Because `tmp` is a pair rather than an iterator, `{ tmp.key(), tmp.value() }` doesn't compile when you actually use the operator. `{ tmp.first, tmp.second }` does compile, but it's returning references to a local variable, which is why its values were sometimes empty and sometimes caused segfaults.

Simple solution to all these problems:
```suggestion
			pair_type operator[](difference_type n) {
				auto tmp = m_iterator + n;
				return { tmp.key(), tmp.value() };
			}
```
Or even simpler:
```suggestion
			pair_type operator[](difference_type n) {
				return *(*this + n);
			}
```